### PR TITLE
[Feature-2815][server] One worker can belong to different workergroups

### PR DESCRIPTION
**Pull request by yangyichao-mango**
## What is the purpose of the pull request

*Implement one worker can belong to different workergroups(#2591 ) (#2815 )*

## Brief change log

  - *Change worker.properties file's `worker.group` conf to `worker.groups`*
  - *Change `server/worker/config/WorkerConfig#workerGroup` to  `server/worker/config/WorkerConfig#workerGroups`*
  - *Change workerRegistry to registry one worker to muti-groups*
  - *Add tests for this issue*

## Verify this pull request

This change added tests and can be verified as follows:

  - *Changed WorkerRegistryTest to verify the change.*

image

### CODE REVIEW

**Comment by yangyichao-mango** [comment]
@dailidong Can your help to review this PR ? Thx.

**Comment by yangyichao-mango** [comment]
> @dailidong Can your help to review this PR ? Thx.

**Comment by dailidong** [comment]
sorry for not review in time, please solve conflicting files first

**Comment by yangyichao-mango** [comment]
Thx a lot for your review @dailidong , I've rebased the branch and resolved the conflicts.

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/registry/WorkerRegistry.java [hunk 1]
```diff
@@ -86,34 +90,39 @@ public void init(){
...
         int workerHeartbeatInterval = workerConfig.getWorkerHeartbeatInterval();
-        this.heartBeatExecutor.scheduleAtFixedRate(new HeartBeatTask(), workerHeartbeatInterval, workerHeartbeatInterval, TimeUnit.SECONDS);
-        logger.info("worker node : {} registry to ZK successfully with heartBeatInterval : {}s", address, workerHeartbeatInterval);
 
+        for (String workerZKPath : workerZkPaths) {
+            zookeeperRegistryCenter.getZookeeperCachedOperator().persistEphemeral(workerZKPath, "");
+            zookeeperRegistryCenter.getZookeeperCachedOperator().getZkClient().getConnectionStateListenable().addListener(new ConnectionStateListener() {
+                @Override
+                public void stateChanged(CuratorFramework client, ConnectionState newState) {
+                    if (newState == ConnectionState.LOST) {
```
Comment (CalvinKirs) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/registry/WorkerRegistry.java:101]: Hello, is it possible to return directly here without continuing to judge? In addition, it is not very intuitive to flood with too many if else

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/registry/WorkerRegistry.java [hunk 2]
```diff
@@ -86,34 +90,39 @@ public void init(){
...
+                        logger.error("worker : {} connection lost from zookeeper", address);
+                    } else if (newState == ConnectionState.RECONNECTED) {
+                        logger.info("worker : {} reconnected to zookeeper", address);
+                        zookeeperRegistryCenter.getZookeeperCachedOperator().persistEphemeral(workerZKPath, "");
+                    } else if (newState == ConnectionState.SUSPENDED) {
+                        logger.warn("worker : {} connection SUSPENDED ", address);
+                    }
```
Comment (CalvinKirs) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/registry/WorkerRegistry.java:108]: Hello, is it possible to return directly here without continuing to judge? In addition, it is not very intuitive to flood with too many if else

Comment (yangyichao-mango) [reply dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/registry/WorkerRegistry.java:108]: > Hello, is it possible to return directly here without continuing to judge? In addition, it is not very intuitive to flood with too many if else

Hi, Thx a lot for your review.
Can you describe your opinion about why it need return directly? I think this is an important step for users to debug zk or worker status errors.

Comment (CalvinKirs) [reply dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/registry/WorkerRegistry.java:108]:
```suggestion
                public void stateChanged(CuratorFramework client, ConnectionState newState) {
                    if (newState == ConnectionState.LOST) {
                        logger.error("worker : {} connection lost from zookeeper", address);
                        return;
                    } 
                     if (newState == ConnectionState.RECONNECTED) {
                        logger.info("worker : {} reconnected to zookeeper", address);
                        zookeeperRegistryCenter.getZookeeperCachedOperator().persistEphemeral(workerZKPath, "");
                        return ;
                    } 
                     if (newState == ConnectionState.SUSPENDED) {
                        logger.warn("worker : {} connection SUSPENDED ", address);
                        return;
                    }
```
This is my suggestion, what do you think,

Comment (yangyichao-mango) [reply dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/registry/WorkerRegistry.java:108]: > This is my suggestion, what do you think,

Thx a lot for your review.
I got your point, and can you describe your opinion about what is the benefits of using
 `if (a) { xxx; return;}  if (b) { xxx; return; }` but not `if (a) { xxx; } else if (b) {xxx;}` ?

**Comment by yangyichao-mango** [comment]
> +1 please resove conflicting , Thx

Hi,
Thx a lot for your review, I have resolved the conflicts.

**Pull request merged**